### PR TITLE
Issue #230 Support PKCE for OAuth 2.0/OpenID Connect authentication module

### PR DESCRIPTION
**Pull request by i7a7467**
## Analysis
When OpenAM works as OAuth2 client/OIDC client, pkce parameter is not available.

## Solution
Extend the OAUTH 2.0 / OPENID CONNECT Authentication module so that you can choose to use PKCE.

## I18N
Add the code challenge method setting to admin console.(In English and Japanese)
Update description of code challenge method.(In English and Japanese)
[realm] - [Authentication] - [Modules] - [(OAUTH 2.0 / OPENID CONNECT module name)] - [Code challenge method]

## Testing
- Code challenge method value is not set. (ex. Upgrade from OpenAM 14.0 )
  - Pkce parameter is not included in authz and token request.
- Code challenge method value is "none".
  - Pkce parameter is not included in authz and token request.
- Code challenge method value is "plain".
  - Pkce parameter is included in authz and token request.
    - code_challenge_method=plain
    - code_challenge=xxx
    - code_verifier=xxx
  - Code challenge and code verifier are the same.
- Code challenge method value is "S256".
  - Pkce parameter is included in authz and token request.
    - code_challenge_method=S256
    - code_challenge=xxx
    - code_verifier=yyy
  - Decoded code challenge and code verifier are the same.

## Regression testing
- Authentication succeeds if the code challenge method value is not set.
- Authentication succeeds if the code challenge method value is "S256".

### CODE REVIEW

**Comment by tsujiguchitky** [comment]
Hi @i7a7467 

Thank you for the wonderful work.

Could you please add a copyright to the source code header?

File: openam-authentication/openam-auth-oauth2/src/main/java/org/forgerock/openam/authentication/modules/oauth2/OAuth.java [hunk 1]
```diff
@@ -275,6 +286,9 @@ public int process(Callback[] callbacks, int state) throws LoginException {
                                 + "contained an unexpected value");
                         throw new AuthLoginException(BUNDLE_NAME, "incorrectState", null);
                     }
+                    if (!pkceCodeChallengeMethod.equalsIgnoreCase("none")) {
+                        pkceCodeVerifier = csrfStateToken.getValue(CoreTokenField.STRING_TWO);
```
Comment (tsujiguchitky) [inline openam-authentication/openam-auth-oauth2/src/main/java/org/forgerock/openam/authentication/modules/oauth2/OAuth.java:290]: I think it's better to throw `AuthLoginException` if CSRF token doesn't contain `code_verifier`. Also, the debug log should be output.

File: openam-authentication/openam-auth-oauth2/src/main/resources/amAuthOAuth.xml [hunk 2]
```diff
@@ -516,6 +527,18 @@ PUBLIC "=//iPlanet//Service Management Services (SMS) 1.0 DTD//EN"
                                      type="single" syntax="string" i18nKey="oidc.crypto_context_value" order="3100"
                                      resourceName="openidConnectContextValue">
                     </AttributeSchema>
+                    <AttributeSchema name="org-forgerock-auth-oauth-code-challenge-method-algorithm"
```
Comment (tsujiguchitky) [inline openam-authentication/openam-auth-oauth2/src/main/resources/amAuthOAuth.xml:530]: This feature will be supported from `openam-jp` edition. If I were you, I would rename the settings. (`org-forgerock-auth-oauth-code-challenge-method-algorithm` -> `jp-openam-auth-oauth-code-challenge-method`).

**Comment by i7a7467** [comment]
@tsujiguchitky Thank you for your review comment.
I fixed the review points and completed the test.
Could you confirm it?

File: openam-authentication/openam-auth-oauth2/src/main/java/org/forgerock/openam/authentication/modules/oauth2/OAuth.java (hunk 1 above)
Comment (tsujiguchitky) [reply openam-authentication/openam-auth-oauth2/src/main/java/org/forgerock/openam/authentication/modules/oauth2/OAuth.java:291]: Thanks for adding exception handling. One more thing, need to add `notFoundCodeVerifier` to properties files.

Comment (i7a7467) [reply openam-authentication/openam-auth-oauth2/src/main/java/org/forgerock/openam/authentication/modules/oauth2/OAuth.java:291]: Thanks. I fixed it.

Comment (tsujiguchitky) [reply openam-authentication/openam-auth-oauth2/src/main/java/org/forgerock/openam/authentication/modules/oauth2/OAuth.java:291]: Good, thanks.

**Comment by tsujiguchitky** [comment]
LGTM

**Review by tsujiguchitky — APPROVED** [review]

**Pull request merged**